ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -1846,19 +1877,51 @@ private class MetadataPutChunk extends PutChunk {
     private final TreeMap<Integer, Pair<StoreKey, Long>> indexToChunkIdsAndChunkSizes;
     private int intermediateChunkSize = routerConfig.routerMaxPutChunkSizeBytes;
     private Pair<? extends StoreKey, BlobProperties> firstChunkIdAndProperties = null;
-    private final BlobId reservedMetadataChunkId;
+    private BlobId reservedMetadataChunkId = null;
 
     /**
      * Initialize the MetadataPutChunk.
+     * If {@link RouterConfig#routerReservedMetadataEnabled} is {@code true}, then this will initialize the reserved
+     * metadata chunk id. There are three cases:
+     * 1. If this is a chunked upload, then the reserved metadata chunk id is passed in the blob properties.
+     * 2. If this is a stitched blob request, then the reserved metadata is obtained from the specified chunksToStitch.
+     * 3. If this is a regular upload, then a new reserved metadata chunk id is generated. Note that this reserved metadata
+     *    id will be used only for composite blobs.
+     *
+     * @param chunksToStitch {@link List} of {@link ChunkInfo} objects to stitch for a stitchBlob request.
+     *                       Will be {@code null} for all other put requests.
+     * @throws RouterException In case of any error.
      */
-    MetadataPutChunk() {
+    MetadataPutChunk(List<ChunkInfo> chunksToStitch) throws RouterException {
       indexToChunkIdsAndChunkSizes = new TreeMap<>();
       // metadata blob is in building state.
       state = ChunkState.Building;
-      reservedMetadataChunkId =
-          ClusterMapUtils.reserveMetadataBlobId(partitionClass, Collections.emptyList(), reservedMetadataIdMetrics,
-              clusterMap, passedInBlobProperties.getAccountId(), passedInBlobProperties.getContainerId(),
-              passedInBlobProperties.isEncrypted(), routerConfig);
+      try {
+        validateReservedMetadataChunkId();
+        if (options.isChunkUpload()) {
+          setReservedMetadataChunkId(passedInBlobProperties.getReservedMetadataBlobId());
+        } else if (chunksToStitch == null) {
+          // if it's not chunked upload, and chunksToStitch is null, then it's a regular upload.
+          reservedMetadataChunkId =
+              ClusterMapUtils.reserveMetadataBlobId(partitionClass, Collections.emptyList(), reservedMetadataIdMetrics,
+                  clusterMap, passedInBlobProperties.getAccountId(), passedInBlobProperties.getContainerId(),
+                  passedInBlobProperties.isEncrypted(), routerConfig);
+        } else if (!chunksToStitch.isEmpty()) {
+          // Empty chunksToStitch list will be handled by {@link processChunksToStitch}.
+          setReservedMetadataChunkId(chunksToStitch.get(0).getReservedMetadataId());
```

Comment:
With existing code, for each chunksToStitch[x], do we provide the reserved meta chunk already?